rust/routee-compass-powertrain/src/model/ice_energy_model.rs

```diff
@@ -0,0 +1,137 @@
+        })
+    }
+}
+
+impl TraversalModelService for IceEnergyModel {
+    fn build(
+        &self,
+        _query: &serde_json::Value,
+    ) -> Result<Arc<dyn TraversalModel>, TraversalModelError> {
+        Ok(Arc::new(self.clone()))
+    }
+}
+
+impl TryFrom<&Value> for IceEnergyModel {
+    type Error = TraversalModelError;
+
+    fn try_from(value: &Value) -> Result<Self, Self::Error> {
+        let config: PredictionModelConfig = serde_json::from_value(value.clone()).map_err(|e| {
+            TraversalModelError::BuildError(format!(
+                "failure reading prediction model configuration: {}",
+                e
+            ))
+        })?;
+        let prediction_model = PredictionModelRecord::try_from(&config)?;
+        let ice_model = IceEnergyModel::new(prediction_model)?;
+        Ok(ice_model)
+    }
+}
+
+impl TraversalModel for IceEnergyModel {
+    fn input_features(&self) -> Vec<(String, InputFeature)> {
+        vec![
+            (
+                String::from(fieldname::EDGE_DISTANCE),
+                InputFeature::Distance(Some(
+                    self.prediction_model_record
+                        .speed_unit
+                        .associated_distance_unit(),
+                )),
+            ),
+            (
+                String::from(fieldname::EDGE_SPEED),
+                InputFeature::Speed(Some(self.prediction_model_record.speed_unit)),
+            ),
+            (
+                String::from(fieldname::EDGE_GRADE),
+                InputFeature::Grade(Some(self.prediction_model_record.grade_unit)),
+            ),
+        ]
+    }
+
+    fn output_features(&self) -> Vec<(String, OutputFeature)> {
+        let energy_unit = self
+            .prediction_model_record
+            .energy_rate_unit
+            .associated_energy_unit();
+        vec![
+            (
+                String::from(fieldname::TRIP_ENERGY),
+                OutputFeature::Energy {
+                    energy_unit,
+                    initial: Energy::ZERO,
+                    accumulator: true,
+                },
+            ),
+            (
+                String::from(fieldname::EDGE_ENERGY),
+                OutputFeature::Energy {
+                    energy_unit,
+                    initial: Energy::ZERO,
+                    accumulator: false,
+                },
+            ),
+        ]
+    }
+
+    fn traverse_edge(
+        &self,
+        _trajectory: (&Vertex, &Edge, &Vertex),
+        state: &mut Vec<StateVariable>,
+        state_model: &StateModel,
+    ) -> Result<(), TraversalModelError> {
+        ice_traversal(state, state_model, self.prediction_model_record.clone())
+    }
+
+    fn estimate_traversal(
+        &self,
+        _od: (&Vertex, &Vertex),
+        state: &mut Vec<StateVariable>,
+        state_model: &StateModel,
+    ) -> Result<(), TraversalModelError> {
+        ice_traversal(state, state_model, self.prediction_model_record.clone())
```

Comment:
Why are we doing a standard traversal here? Wouldn't it make more sense to use the minimum energy rate over the distance estimate? If we a standard traversal, we take the speed estimate which gets set to the highest possible speed and for an ICEV we typically see the highest energy consumption at large speeds.

Reply:
you're totally right! i forgot to fix this for ICE after i did fix it for the BEV and PHEV models. on it!